internal/cephfs/controllerserver.go

```diff
@@ -294,3 +444,309 @@ func (cs *ControllerServer) ControllerExpandVolume(ctx context.Context, req *csi
+	if err := cs.validateSnapshotReq(ctx, req); err != nil {
+		return nil, err
+	}
+	cr, err := util.NewAdminCredentials(req.GetSecrets())
+	if err != nil {
+		return nil, err
+	}
+	defer cr.DeleteCredentials()
+
+	clusterData, err := getClusterInformation(req.GetParameters())
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	requestName := req.GetName()
+	sourceVolID := req.GetSourceVolumeId()
+	// Existence and conflict checks
+	if acquired := cs.SnapshotLocks.TryAcquire(requestName); !acquired {
+		klog.Errorf(util.Log(ctx, util.SnapshotOperationAlreadyExistsFmt), requestName)
+		return nil, status.Errorf(codes.Aborted, util.SnapshotOperationAlreadyExistsFmt, requestName)
+	}
+	defer cs.SnapshotLocks.Release(requestName)
+
+	if err = cs.OperationLocks.GetSnapshotCreateLock(sourceVolID); err != nil {
+		klog.Error(util.Log(ctx, err.Error()))
+		return nil, status.Error(codes.Aborted, err.Error())
+	}
+
+	defer cs.OperationLocks.ReleaseSnapshotCreateLock(sourceVolID)
+
+	// Find the volume using the provided VolumeID
+	parentVolOptions, vid, err := newVolumeOptionsFromVolID(ctx, sourceVolID, nil, req.GetSecrets())
+	if err != nil {
+		if errors.Is(err, util.ErrPoolNotFound) {
+			klog.Warningf(util.Log(ctx, "failed to get backend volume for %s: %v"), sourceVolID, err)
+			return nil, status.Error(codes.NotFound, err.Error())
+		}
+
+		if errors.Is(err, ErrVolumeNotFound) {
+			return nil, status.Error(codes.NotFound, err.Error())
+		}
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if clusterData.ClusterID != parentVolOptions.ClusterID {
+		return nil, status.Errorf(codes.InvalidArgument, "requested cluster id %s not matching subvolume cluster id %s", clusterData.ClusterID, parentVolOptions.ClusterID)
+	}
+
+	cephfsSnap := genSnapFromOptions(ctx, req)
+	if cephfsSnap == nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// lock out parallel snapshot create operations
+	if acquired := cs.VolumeLocks.TryAcquire(sourceVolID); !acquired {
+		klog.Errorf(util.Log(ctx, util.VolumeOperationAlreadyExistsFmt), sourceVolID)
+		return nil, status.Errorf(codes.Aborted, util.VolumeOperationAlreadyExistsFmt, sourceVolID)
+	}
+	defer cs.VolumeLocks.Release(sourceVolID)
+	snapName := req.GetName()
+	sid, snapInfo, err := checkSnapExists(ctx, parentVolOptions, vid.FsSubvolName, cephfsSnap, cr)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	// check are we able to retrieve the size of parent
+	// ceph fs subvolume info command got added in 14.2.10 and 15.+
+	// as we are not able to retrieve the parent size we are rejecting the
+	// request to create snapshot.
+	// TODO: For this purpose we could make use of cached clusterAdditionalInfo too.
+	var info Subvolume
+	info, err = getSubVolumeInfo(ctx, parentVolOptions, cr, volumeID(vid.FsSubvolName))
+	if err != nil {
+		// We could either use ClusterAdditionInfo[volOptions.ClusterID].subVolumeInfoSupported
```

Comment:
This will not work for the first the `subVolumeInfoSupported` will be false. we will always return the error to the user.

Reply:
got rid of cache config completely as in other review comment.